### ISSUE TITLE
Missing samples when creating phyloseq object

### ISSUE THREAD

**Issue by Jasmine84** [joey711/phyloseq#1208]
Hello

When I create my phyloseq object, no error occurs. However, of the 125 samples only 95 samples are in my phyloseq object. Both otu_table and metadata contains 125 samples. Can you help with this issue?

# There are 125 samples
dim(metadata_secondrun) # 125 samples 
rownames(seqtab.nochim) # 125 samples

ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), 
               sample_data(metadata_secondrun),
               tax_table(taxa))
ps
sample_names(ps)
# ONLY 95 samples matches and we should have 125 samples

Best regards, Jasmine

**Comment by evertonjlima** [comment]
Have you checked your sample data, before adding it to the phyloseq object? If I am not mistaken, samples for which the metadat is no present in will get removed.

**Comment by Jasmine84** [comment]
Thanks for anwering @evertonjlima! :) There are 125 samples in my metadata and it looks like the names are the same in the otu_table. However I do not know if you also have to check the taxa table (do not know how to do this). Do you know a R-code for checking it in R that the sample names are the same in otu_table and metadata and if necessary taxa table?

**Comment by evertonjlima** [comment]
I don't believe checking the taxa table is necessary. In any case, you can do so by getting the taxa from your phyloseq object and comparing your raw input (seqtab.nochim).


Maybe this can help you get started?
`
tax_tb <- tax_table(ps)
all(tax_tb %in% colnames(seqtab.nochim)[-1] )
`

**Comment by Jasmine84** [comment]
I found a solution to easy check the samples IDs in OTU table and sampledata:

vector1 <- rownames(sampledata)
vector2 <- rownames(otu_table)
vector1 == vector2

 So problem solved! Thanks for the help @evertonjlima

**Comment by jeaninebr** [comment]
Hello Github world. 

I am having the same issue as @Jasmine84 :

I have 384 samples from a qime pipeline. I checked my sample_file and my otu_table, there are 384 samples in both of them, the names are the identical. 
I then try to import them with the command 'import_qime' and in the phyloseq object, only 380 samples are present. I don't get an error or warning !

So here is my code so far: 

otufile     <- "~/e_OTU-3/p677_run200221_COI_ZOTU_c99_Count_Sintax.txt"
mapfile    <-  "~/e_OTU-3/p677_JB_Co1_200323_Mapfile.txt"
data_raw <- import_qiime(otufilename = otufile, mapfilename = mapfile)

So my first guess was that the names are not the same (@evertonjlima ), but when I import the files and I use setdiff() on the names, no entry is returned. When I use intersect(), I get the full list of 384 samples names. 

Also my names start with a letter, the four samples dropped are: 
T152VS-L2, 
T152VS-R2, 
T021VD-L2, 
T021VD-R2.  
Funnily, I have replicates of these and they are imported without any problem, those are named: T152VS-L1, 
T152VS-R1, 
T021VD-L1, 
T021VD-R1. 

So really, I have no idea what else could be wrong. I work around now, by manually replacing the otu_table from the individually imported otu_file, but I'd still want to know, what the problem is. 
I also restarted R , removed the objects, reloaded phyloseq, but that didn't help either. Any suggestions?

**Comment by wlwhalley** [comment]
Hi, 

I am also having the same issue of phyloseq mysteriously dropping samples. 

I used rds files form the dada2 pipelines such as this:

```
fitGTR<-readRDS("C:/VMshared/fungi/analysis/treeits_prefilt.rds")
seqtab<-readRDS("C:/VMshared/fungi/analysis/seqtab_ITSpair_final1.rds")
tax<-readRDS("C:/VMshared/fungi/analysis/tax_ITSpair_final1.rds")
metadata_ITS <- read.csv("C:/VMshared/fungi/ph data/masterph.csv", row.names=1)
tax<-as.matrix(tax)  ## <- this is required
phy <- phyloseq(otu_table(seqtab, taxa_are_rows=FALSE), 
               sample_data(metadata_ITS), 
               tax_table(tax),phy_tree(fitGTR$tree))
phy

seqtab contains the samples that are dropped. 

I also did this on a seperate set of data, with the same sample names (bacteria and fungi) and the exact same sample names were dropped. Very frustrating. 

The sample names were:
P5Y1
P5Y2
P5Y3
E5Y1
E5Y2
E5Y3

But the following were not dropped for some reason
S5Y1
S5Y2
S5Y3
```

**Comment by wlwhalley** [comment]
Embarrassingly my sample names did not match, which is not much help to you. 

Do you check the structure after importing as a text file? I checked three times but didn't notice the slightly different sample names until the fourth check

**Comment by jeaninebr** [comment]
Thanks for your responses. I did not figure it out. I opened the raw files
for OTUs and samples and mapfile, ordered the samples alphabetically and
copied it from the one table to the others to make sure that the names are
identical (e.g. even if there is a typo or similar it would be the same in
all of them). still no success.

I ended up creating a phyloseq object with less samples than expected and
overwriting the individual components of the created object. Here is the
code for my workaround. Hope this helps...

## --------------------------------
## Data Import in R
## --------------------------------
## File containing OTU vs. sample site
otufile     <- "~/Desktop/Leese/p677_run201029_Count_Sintax.txt"
## File containing metadata
mapfile    <-  "~/Desktop/Mapfile.txt"
## File containing sequence of each OTU
refseq      <- "~/Desktop/p677_run201029.afa"
## File containing phylogenetic tree
treefile    <- "~/Desktop/p677_run201029_CLU.tre"
############################################################
## importing
############################################################
## qime format
data_L   <- import_qiime(otufilename = otufile, mapfilename = mapfile,
treefilename = treefile, refseqfilename = refseq)

#### HERE THE SAMPLES WERE DROPPED, so:

## convert it into otu_table
otu2 <- data.frame(otu)
## add rows names
rownames(otu2) <- otu2[,1]
# remove OTU name as they are now rownames (first column)
otu2 <-otu2[,-1]
## remove Syntax column (last column with the taxonomic assignment) from
OTU table
otu2 <-otu2[,-(385)]
## add column names from the mapfile, check order of samples!
colnames(otu2) <- sort(mapf$SampleID)
## convert into matrix and then otu_table
otu3 <- as.matrix(otu2)
otu3
OTU <-  otu_table(otu3, taxa_are_rows = TRUE)
## convert into sample_data
row.names(mapf) <- sort(mapf$SampleID)
MAP <- sample_data(mapf)
## HERE I OVERWRITE THE TWO OBJECTS otu_table and mapfile.
***@***.***_table <- OTU
***@***.***_data <- MAP





Am Di., 8. Juni 2021 um 22:18 Uhr schrieb wlwhalley <
***@***.***>:

**Comment by jeaninebr** [comment]
Sorry, forgot two lines of code, here we go again:


## --------------------------------
## Data Import in R
## --------------------------------
## File containing OTU vs. sample site
otufile     <- "~/Desktop/Leese/p677_run201029_Count_Sintax.txt"
## File containing metadata
mapfile    <-  "~/Desktop/Mapfile.txt"
## File containing sequence of each OTU
refseq      <- "~/Desktop/p677_run201029.afa"
## File containing phylogenetic tree
treefile    <- "~/Desktop/p677_run201029_CLU.tre"
############################################################
## importing
############################################################
## qime format
data_L   <- import_qiime(otufilename = otufile, mapfilename = mapfile,
treefilename = treefile, refseqfilename = refseq)

####In this object THE SAMPLES WERE DROPPED, so:

## reload OTU
otu <- read_delim("~/Desktop/Leese/p677_run201029_Count_Sintax.txt",
                  "\t", escape_double = FALSE, trim_ws = TRUE)

## reload mapfile
mapf <- read.delim("~/Desktop/Mapfile.txt")

## convert it into otu_table
otu2 <- data.frame(otu)
## add rows names
rownames(otu2) <- otu2[,1]
# remove OTU name as they are now rownames (first column)
otu2 <-otu2[,-1]
## remove Syntax column (last column with the taxonomic assignment) from
OTU table
otu2 <-otu2[,-(385)]
## add column names from the mapfile, check order of samples!
colnames(otu2) <- sort(mapf$SampleID)
## convert into matrix and then otu_table
otu3 <- as.matrix(otu2)
otu3
OTU <-  otu_table(otu3, taxa_are_rows = TRUE)
## convert into sample_data
row.names(mapf) <- sort(mapf$SampleID)
MAP <- sample_data(mapf)
## HERE I OVERWRITE THE TWO OBJECTS otu_table and mapfile.
***@***.***_table <- OTU
***@***.***_data <- MAP


Am Do., 10. Juni 2021 um 23:10 Uhr schrieb Jeanine Brantschen <
***@***.***>:

**Comment by science-chump** [comment]
Hey, assuming you you all have moved on from this issue but for future reference I found a possible solution to this as I was running into the same problem.

For me it was the presence of having hyphens. I simply replaced all my "-" with "." in excel and it worked fine. Guessing that either R or phyloseq doesn't appreciate the hyphens.

**Comment by jeaninebr** [comment]
Hey there.

I now worked around by loading the two tables (otu and tax) as .csv files,
and then assigning them the proper class manually, and then building a
phyloseq from scratch, adding the tables manually.

I don't know what the actual issue is, but this worked for me.

Am Do., 21. Jan. 2021 um 12:10 Uhr schrieb wlwhalley <
***@***.***>:

**Comment by FabriceBouilloux** [comment]
I got the same kind of issue. The origin of the problem was the names of samples. The missing sample names contained a "-" character. I replace it by "_" and it fix it.
